### PR TITLE
[WIP] r/aws_dynamodb_table: Support 11/15/2018 Server-Side Encryption enhancements

### PR DESCRIPTION
**Pull request by ewbankkit**
Fixes https://github.com/terraform-providers/terraform-provider-aws/issues/5637.

Currently _WIP_:

```console
$ terraform apply
aws_dynamodb_table.basic-dynamodb-table: Refreshing state... (ID: test-enc-01)
aws_dynamodb_table.basic-dynamodb-table: Modifying... (ID: test-enc-01)
  server_side_encryption.0.enabled:  "false" => "true"
  server_side_encryption.0.sse_type: "" => "AES256"
Error applying plan:

1 error(s) occurred:

* aws_dynamodb_table.basic-dynamodb-table: 1 error(s) occurred:

* aws_dynamodb_table.basic-dynamodb-table: ValidationException: One or more parameter values were invalid: Update table encryption is coming to your account soon
	status code: 400, request id: BTJKPO1DQ5RMBQ6TDT4IT85LDVVV4KQNSO5AEMVJF66Q9ASUAAJG

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

Still TODO:
- [x] Update `aws_dynamodb_table` data source
- [x] More acceptance tests

### CODE REVIEW

**Comment by bflad** [comment]
@ewbankkit any chance this could be updated? Thanks!

**Comment by ewbankkit** [comment]
@bflad Rebased to remove conflicts. The service changes have not yet been released:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDynamoDbTable_encryption'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSDynamoDbTable_encryption -timeout 120m
=== RUN   TestAccAWSDynamoDbTable_encryption
--- FAIL: TestAccAWSDynamoDbTable_encryption (4.49s)
    testing.go:527: Step 0 error: Error applying: 1 error occurred:
        	* aws_dynamodb_table.basic-dynamodb-table: 1 error occurred:
        	* aws_dynamodb_table.basic-dynamodb-table: ValidationException: One or more parameter values were invalid: Unsupported SSEType AES256
        	status code: 400, request id: 9S3LR33ADF60MHAKRG598KMAL3VV4KQNSO5AEMVJF66Q9ASUAAJG
        
        
        
        
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	4.500s
GNUmakefile:20: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

**Comment by ewbankkit** [comment]
Yes, I think [this](https://aws.amazon.com/about-aws/whats-new/2018/11/amazon-dynamodb-encrypts-all-customer-data-at-rest/) is the announcement. I will try and finish this today.

**Comment by ewbankkit** [comment]
Running the tests unmodified I still get the same error:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDynamoDbTable_encryption'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAWSDynamoDbTable_encryption -timeout 120m
=== RUN   TestAccAWSDynamoDbTable_encryption
=== PAUSE TestAccAWSDynamoDbTable_encryption
=== CONT  TestAccAWSDynamoDbTable_encryption
--- FAIL: TestAccAWSDynamoDbTable_encryption (4.39s)
    testing.go:538: Step 0 error: Error applying: 1 error occurred:
        	* aws_dynamodb_table.basic-dynamodb-table: 1 error occurred:
        	* aws_dynamodb_table.basic-dynamodb-table: ValidationException: One or more parameter values were invalid: SSEType AES256 is not supported
        	status code: 400, request id: SGMHVUIPQ8LUDLL6M87JRTKM8VVV4KQNSO5AEMVJF66Q9ASUAAJG
        
        
        
        
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	4.398s
GNUmakefile:20: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

and if I change the default `sse_type` from `AES256` to `KMS` I get a different error:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDynamoDbTable_encryption'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAWSDynamoDbTable_encryption -timeout 120m
=== RUN   TestAccAWSDynamoDbTable_encryption
=== PAUSE TestAccAWSDynamoDbTable_encryption
=== CONT  TestAccAWSDynamoDbTable_encryption
--- FAIL: TestAccAWSDynamoDbTable_encryption (19.66s)
    testing.go:538: Step 0 error: Error applying: 1 error occurred:
        	* aws_dynamodb_table.basic-dynamodb-table: 1 error occurred:
        	* aws_dynamodb_table.basic-dynamodb-table: ValidationException: One or more parameter values were invalid: Update table encryption is coming to your account soon
        	status code: 400, request id: PIH23LT13T3FDGF0MD9LDA5OS7VV4KQNSO5AEMVJF66Q9ASUAAJG
        
        
        
        
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	19.679s
GNUmakefile:20: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

Now, suppose I create a new table via the console and accept all defaults. I see

tf-test-default-encryption

and at the API level:
```console
$ aws --region us-east-1 dynamodb describe-table --table-name tf-test-default-encryption
{
    "Table": {
        "TableArn": "arn:aws:dynamodb:us-east-1:000000000000:table/tf-test-default-encryption", 
        "AttributeDefinitions": [
            {
                "AttributeName": "name", 
                "AttributeType": "S"
            }
        ], 
        "ProvisionedThroughput": {
            "NumberOfDecreasesToday": 0, 
            "WriteCapacityUnits": 5, 
            "ReadCapacityUnits": 5
        }, 
        "TableSizeBytes": 0, 
        "TableName": "tf-test-default-encryption", 
        "TableStatus": "ACTIVE", 
        "TableId": "965e2d09-2e0a-44c4-a604-9da95ff08547", 
        "KeySchema": [
            {
                "KeyType": "HASH", 
                "AttributeName": "name"
            }
        ], 
        "ItemCount": 0, 
        "CreationDateTime": 1542379431.603
    }
}
```

If I create a new table via the console and choose KMS encryption, I see

tf-test-explicit-encryption

and at the API level:

```console
$ aws --region us-east-1 dynamodb describe-table --table-name tf-test-explicit-encryption
{
    "Table": {
        "TableArn": "arn:aws:dynamodb:us-east-1:000000000000:table/tf-test-explicit-encryption", 
        "AttributeDefinitions": [
            {
                "AttributeName": "name", 
                "AttributeType": "S"
            }
        ], 
        "ProvisionedThroughput": {
            "NumberOfDecreasesToday": 0, 
            "WriteCapacityUnits": 5, 
            "ReadCapacityUnits": 5
        }, 
        "TableSizeBytes": 0, 
        "TableName": "tf-test-explicit-encryption", 
        "TableStatus": "ACTIVE", 
        "TableId": "eeee1609-5ad3-4e28-a28e-ce80736e95f9", 
        "SSEDescription": {
            "Status": "ENABLED", 
            "KMSMasterKeyArn": "arn:aws:kms:us-east-1:000000000000:key/2c42622a-8fb4-4147-b98c-8dac4aa33bd4", 
            "SSEType": "KMS"
        }, 
        "KeySchema": [
            {
                "KeyType": "HASH", 
                "AttributeName": "name"
            }
        ], 
        "ItemCount": 0, 
        "CreationDateTime": 1542379472.602
    }
}
```

**Comment by ewbankkit** [comment]
Re-scoped this PR to implement the functionality available today.
Omitting `server_side_encryption` or specifying

```hcl
server_side_encryption {
  enabled = false
}
```

creates a DynamoDB table with _DEFAULT_ encryption - "AWS owned CMK".

Specifying 

```hcl
server_side_encryption {
  enabled = false
}
```

or

```hcl
server_side_encryption {
  enabled  = false
  sse_type = "KMS"
}
```

or even

```hcl
server_side_encryption {
  enabled           = false
  sse_type          = "KMS"
  kms_master_key_id = "alias/aws/dynamodb"
}
```

creates a DynamoDB table with _KMS_encryption - ["AWS managed CMK"](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/encryption.howitworks.html#service-default-kms).

File: aws/data_source_aws_dynamodb_table.go [hunk 1]
```diff
@@ -15,14 +15,30 @@ func dataSourceAwsDynamoDbTable() *schema.Resource {
 	return &schema.Resource{
 		Read: dataSourceAwsDynamoDbTableRead,
 		Schema: map[string]*schema.Schema{
+			"arn": {
```
Comment (ewbankkit) [inline aws/data_source_aws_dynamodb_table.go]: Reordered the `data_source_aws_dynamodb_table` attributes to match  `resource_aws_dynamodb_table` for easier checking of missing attributes.

File: aws/data_source_aws_dynamodb_table.go [hunk 2]
```diff
@@ -145,31 +158,38 @@ func dataSourceAwsDynamoDbTable() *schema.Resource {
...
 					},
 				},
 			},
-			"write_capacity": {
-				Type:     schema.TypeInt,
-				Computed: true,
-			},
-			"server_side_encryption": {
+			"tags": tagsSchemaComputed(),
+			"point_in_time_recovery": {
```
Comment (ewbankkit) [inline aws/data_source_aws_dynamodb_table.go]: Added `point_in_time_recovery` as the documentation says that the data source returns all the same attributes as the resource.

**Comment by ewbankkit** [comment]
As-is this PR does not address any of the enhancements listed in the original issue's [first comment](https://github.com/terraform-providers/terraform-provider-aws/issues/5637#issue-352692407) but will form a foundation for when those features are available.

**Comment by ewbankkit** [comment]
Acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccDataSourceAwsDynamoDbTable_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccDataSourceAwsDynamoDbTable_ -timeout 120m
=== RUN   TestAccDataSourceAwsDynamoDbTable_basic
=== PAUSE TestAccDataSourceAwsDynamoDbTable_basic
=== CONT  TestAccDataSourceAwsDynamoDbTable_basic
--- PASS: TestAccDataSourceAwsDynamoDbTable_basic (136.68s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	136.689s

$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDynamoDbTable_encryption'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAWSDynamoDbTable_encryption -timeout 120m
=== RUN   TestAccAWSDynamoDbTable_encryptionDefault
=== PAUSE TestAccAWSDynamoDbTable_encryptionDefault
=== CONT  TestAccAWSDynamoDbTable_encryptionDefault
--- PASS: TestAccAWSDynamoDbTable_encryptionDefault (121.69s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	121.717s
```

**Comment by jpalomaki** [comment]
FWIW, I just created a DynamoDB table, specifying:
```
server_side_encryption {
    enabled = true
}
```
which created the table with encryption type `KMS`, while I specifically hoped to get `DEFAULT`.

So at least I see value in this, being able to distinguish between the two explicitly.

**Comment by ewbankkit** [comment]
Closed. Replaced by https://github.com/terraform-providers/terraform-provider-aws/pull/7499.

**Comment by ewbankkit** [comment]
@jpalomaki This all boils down to a documentation update.
To get `DEFAULT` encryption you should specify `enabled = false` in the `server_side_encryption` block (or leave the whole block out). This is kind of weird but was the way that AWS implemented [encrypt everything](https://aws.amazon.com/about-aws/whats-new/2018/11/amazon-dynamodb-encrypts-all-customer-data-at-rest/) in the [API](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_SSESpecification.html).

**Pull request closed without merging**